crates/fluvio-protocol-derive/src/util.rs

```diff
@@ -1,75 +1,256 @@
-        match &attr.lit {
-            Lit::Int(version_val) => {
-                //  println!("version value: {}",version_val.value());
-                version_val.base10_parse::<u64>().ok()
+pub fn get_lit_int_value<'a>(
+    attr_name: &'a str,
+    value: &'a Option<Expr>,
+    span: Span,
+) -> syn::Result<i16> {
+    if let Some(value_expr) = value {
+        parse_int_expr(attr_name, value_expr, span)
+    } else {
+        Err(syn::Error::new(
+            span,
+            format!("Expected {attr_name} to be valid Expr: `{attr_name} = \"...\"`"),
+        ))
+    }
+}
+pub fn get_lit_int<'a>(
+    attr_name: &'a str,
+    value: &'a Option<Expr>,
+    span: Span,
+) -> syn::Result<&'a syn::LitInt> {
+    match &value {
+        Some(Expr::Lit(lit_expr)) => {
+            if let Lit::Int(lit) = &lit_expr.lit {
+                Ok(lit)
+            } else {
+                Err(syn::Error::new(
+                    span,
+                    format!("Expected {attr_name} attribute to be an int: `{attr_name} = \"...\"`"),
+                ))
             }
-            _ => unimplemented!(),
         }
-    } else {
-        None
+        _ => Err(syn::Error::new(
+            span,
+            format!("Expected {attr_name} attribute to be a Lit: `{attr_name} = \"...\"`"),
+        )),
+    }
+}
+pub fn get_lit_str<'a>(
+    attr_name: &'a str,
+    value: &'a Option<Expr>,
+    span: Span,
+) -> syn::Result<&'a syn::LitStr> {
+    match &value {
+        Some(Expr::Lit(lit_expr)) => {
+            if let Lit::Str(lit) = &lit_expr.lit {
+                Ok(lit)
+            } else {
+                Err(syn::Error::new(
+                    span,
+                    format!(
+                        "Expected {attr_name} attribute to be a string: `{attr_name} = \"...\"`"
+                    ),
+                ))
+            }
+        }
+        _ => Err(syn::Error::new(
+            span,
+            format!("Expected {attr_name} attribute to be a Lit: `{attr_name} = \"...\"`"),
+        )),
     }
 }
 
-/// find name value with str value
-pub(crate) fn find_string_name_value(version_meta: &Meta, attr_name: &str) -> Option<LitStr> {
-    if let Some(attr) = find_name_attribute(version_meta, attr_name) {
-        match &attr.lit {
-            Lit::Str(val) => Some(val.clone()),
-            _ => unimplemented!(),
+pub(crate) fn find_attr(attrs: &[Attribute], name: &str) -> Option<Meta> {
+    attrs.iter().find_map(|a| {
+        if a.meta.path().is_ident(name) {
+            Some(a.meta.clone())
+        } else {
+            None
         }
-    } else {
-        None
+    })
+}
+pub(crate) fn find_name_value_from_meta_list(
+    meta_list: &MetaList,
+    attr_name: &str,
+) -> Option<MetaNameValue> {
+    for item in meta_list
+        .parse_args_with(Punctuated::<syn::Meta, Token![,]>::parse_terminated)
+        .unwrap_or_else(|err| panic!("Invalid #[new] attribute: {}", err))
```

Comment:
We should use error with the proper span instead of panicking

Reply:
I changed it so it returns **Result** instead of **Option** on most of the places and improved on the error handling.